### PR TITLE
feat: NIXL Based RDMA Support w/ Multimodal Example

### PR DESCRIPTION
**Pull request by whoisj**
This PR introduces a new component library for Dynamo which supports RDMA based read/writes between GPU allocations using the NIXL library.

Two modes of operation are supported:

1. RECV side creates a `WritableOperation` to which the SEND side will pair a `WriteOperation` to complete the RDMA transaction.
2. SEND side creates a `ReadableOperation` to which the RECV side will pair a `ReadOperation` to complete the RDMA transaction.
    
Additionally, the library provide a more "pythanic" style and method than the base NIXL Python wheel provides.

The example itself was updated to provide NIXL support via the dynamo.connect library for moving outputs from one worker to the inputs of another.
    
This change includes renaming (`VllmWorker` => `VllmDecodeWorker`, `EncodeWorker` => `VllmEncodeWorker`, and `PrefillWorker` => `VllmPrefillWorker`) to provide consistency and clarity.

I recommend reviewers start with the decode, encode, and prefill workers.

### CODE REVIEW

**Comment by whoisj** [comment]
After 300 minor changes to appease the linter gods, this PR is actually ready for review.

File: examples/multimodal/components/decode_worker.py [hunk 1]
```diff
@@ -141,25 +142,42 @@ async def async_init(self):
                 vision_tower.vision_model.embeddings.position_embedding.num_embeddings
             )
         else:
-            enc_comp_ns, enc_comp_name = EncodeWorker.dynamo_address()  # type: ignore
+            EMBEDDINGS_SHAPE = (1, 577, 4096)
```
Comment (krishung5) [inline examples/multimodal/components/decode_worker.py:145]: Instead of hardcoded here, could we use something like line 142 to get the shape
```
self.embedding_size = (
                vision_tower.vision_model.embeddings.position_embedding.num_embeddings
            )
```

File: examples/multimodal/components/prefill_worker.py [hunk 2]
```diff
@@ -38,6 +39,11 @@
 
 logger = logging.getLogger(__name__)
 
+# Constants for the shape and dtype of the embeddings tensor.
+EMBEDDINGS_SHAPE = (1, 577, 4096)
```
Comment (krishung5) [inline examples/multimodal/components/prefill_worker.py:43]: Same here, could we remove the hardcoded shape?

File: examples/multimodal/connect/__init__.py [hunk 3]
```diff
@@ -0,0 +1,1471 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (krishung5) [inline examples/multimodal/connect/__init__.py:1]: Should we keep the __init__.py file empty? Also I think the license section is missing a couple of lines.

File: examples/multimodal/components/decode_worker.py (hunk 1 above)
Comment (whoisj) [reply examples/multimodal/components/decode_worker.py:145]: makes sense, once I get my test machine allocation back I'll make and verify the change.

File: examples/multimodal/connect/__init__.py (hunk 3 above)
Comment (whoisj) [reply examples/multimodal/connect/__init__.py:1]: @nnshah1 seemed to think this was fine. There were huge issues when breaking the code out into separate files.

File: examples/multimodal/components/decode_worker.py (hunk 1 above)
Comment (whoisj) [reply examples/multimodal/components/decode_worker.py:145]: `vision_tower` is not defined in the scope; so it doesn't seem like this will work.

File: examples/multimodal/components/prefill_worker.py (hunk 2 above)
Comment (whoisj) [reply examples/multimodal/components/prefill_worker.py:43]: again, `vision_tower` is not defined in scope here either.

File: examples/multimodal/connect/__init__.py (hunk 3 above)
Comment (whoisj) [reply examples/multimodal/connect/__init__.py:1]: re: copyright, the checker is happy: 

> [PASS] examples/multimodal/connect/__init__.py

File: examples/multimodal/components/decode_worker.py (hunk 1 above)
Comment (krishung5) [reply examples/multimodal/components/decode_worker.py:145]: Maybe we could make the vision_tower one layer before so that it could be referred to both if-else section?

Comment (whoisj) [reply examples/multimodal/components/decode_worker.py:145]: but the model doesn't need to be loaded in both cases, right? isn't that just wasting GPU resources?

Comment (krishung5) [reply examples/multimodal/components/decode_worker.py:145]: That's a good concern, maybe we could unload the model? Just wanted to see if we could get rid of the hardcoded stuff and infer that info somehow.

File: examples/multimodal/connect/__init__.py [hunk 4]
```diff
@@ -0,0 +1,1471 @@
...
+        return self
+
+    def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
+        self.__release__()
+
+    def __release__(self) -> None:
+        """
+        Private method to release resources. Only to be called by `self`.
+        """
+        pass
```
Comment (rmccorm4) [inline examples/multimodal/connect/__init__.py:128]: (1) ~~Is this supposed to be releasing some resources?~~ EDIT: I see this is just the base class, ignore

(2) Idiomatic Python-wise, [dunder methods](https://www.geeksforgeeks.org/dunder-magic-methods-python/) `__xyz__` are usually only used for well-defined built-in concepts that has some meaning or get automatically called by other built-ins (like `__str__`/`__repr__` for `str()`/`print()`, or `__len__`/`__next__` for iterators/lists/etc.). 
 
If this is a custom method just intended to be private, I would typically just do the `_` prefix `_release`. I don't feel strongly about saying this must be changed here, just sharing that it's a little uncommon to do this.

File: examples/multimodal/components/decode_worker.py [hunk 5]
```diff
@@ -141,25 +142,42 @@ async def async_init(self):
...
                 .component(enc_comp_name)
                 .endpoint("encode")
                 .client()
             )
 
+            self._connector = connect.Connector(runtime=runtime, namespace=enc_comp_ns)
+            await self._connector.initialize()
+
+            # Create a longer-lived buffer for receiving the image embeddings.
+            embeddings = torch.empty(
```
Comment (ptarasiewiczNV) [inline examples/multimodal/components/decode_worker.py:161]: Will it work with bs > 1? Or do we need to process images one by one because we have fixed embeddings buffer that encode_worker writes into?

File: examples/multimodal/components/encode_worker.py [hunk 6]
```diff
@@ -50,32 +68,153 @@ def __init__(self) -> None:
...
+        # 8. Yield the encode response.
+
+        # Either retrieve the image from the cache or download it and then cache it.
+        if request.image_url in self._image_cache:
+            image = self._image_cache[request.image_url]
+            logger.debug(
+                f"Image found in cache for request: {{ id: {request.request_id}, image_url: '{request.image_url}' }}."
+            )
+        else:
+            image = self.open_image(request.image_url)
```
Comment (ptarasiewiczNV) [inline examples/multimodal/components/encode_worker.py]: Does it make sense to download images asynchronously?

File: examples/multimodal/connect/__init__.py (hunk 3 above)
Comment (ptarasiewiczNV) [reply examples/multimodal/connect/__init__.py:1]: I support the idea of splitting this module into smaller ones and leaving init mainly for imports. What is the issue with that?

Comment (ptarasiewiczNV) [inline examples/multimodal/connect/__init__.py:1]: This is basically NIXL higher level API on top of the agent. I think at some point we could consider moving this into NIXL python API as it probably will be a good entrypoint for a lot of users. CC @mkhazraee

Comment (whoisj) [reply examples/multimodal/connect/__init__.py:1]: My goal was to make NIXL easier to use. The current Python API isn't trivial to understand without assistance.

File: examples/multimodal/connect/__init__.py (hunk 4 above)
Comment (whoisj) [reply examples/multimodal/connect/__init__.py:128]: fixed up.

File: examples/multimodal/components/decode_worker.py (hunk 5 above)
Comment (whoisj) [reply examples/multimodal/components/decode_worker.py:161]: no idea. @krishung5, any thoughts here?

File: examples/multimodal/components/encode_worker.py (hunk 6 above)
Comment (whoisj) [reply examples/multimodal/components/encode_worker.py]: maybe? This PR is about adding the NIXL support, not so much about replacing the existing code everywhere.

@krishung5 , thoughts?

Comment (piotrm-nvidia) [reply examples/multimodal/components/encode_worker.py]: The image download time is 0.6 second for Coco dataset so download takes longer than processing, which is 0.3 second for 8B llava vLLM baseline.

We can keep this as TODO and assume few milliseconds download time for PoC purpose.

Comment (piotrm-nvidia) [reply examples/multimodal/components/encode_worker.py]: I don't see any asynchronous download of images in vLLM


https://github.com/vllm-project/vllm/blob/43b5f61dceb4ec1ac17e08d847162cd86344aba1/vllm/entrypoints/chat_utils.py#L709C1-L739C43

```python
class MultiModalContentParser(BaseMultiModalContentParser):

    def __init__(self, tracker: MultiModalItemTracker) -> None:
        super().__init__()

        self._tracker = tracker

        self._connector = MediaConnector(
            allowed_local_media_path=tracker.allowed_local_media_path,
        )

    def parse_image(self, image_url: str) -> None:
        image = self._connector.fetch_image(image_url)

        placeholder = self._tracker.add("image", image)
        self._add_placeholder(placeholder)

    def parse_image_embeds(self,
                           image_embeds: Union[str, dict[str, str]]) -> None:
        if isinstance(image_embeds, dict):
            embeds = {
                k: self._connector.fetch_image_embedding(v)
                for k, v in image_embeds.items()
            }
            placeholder = self._tracker.add("image_embeds", embeds)

        if isinstance(image_embeds, str):
            embedding = self._connector.fetch_image_embedding(image_embeds)
            placeholder = self._tracker.add("image_embeds", embedding)

        self._add_placeholder(placeholder)
``` 

load_from_url implementation:

https://github.com/vllm-project/vllm/blob/43b5f61dceb4ec1ac17e08d847162cd86344aba1/vllm/multimodal/utils.py#L93

File: examples/multimodal/connect/__init__.py (hunk 3 above)
Comment (mkhazraee) [reply examples/multimodal/connect/__init__.py:1]: Piotr and I talked the other day, check_remote_xfer_done is the new API that J is using and useful for users. Other than that, the communication is still async 1-sided, so the APIs from NIXL are there. Let's not complicate NIXL side with the same functionality, just a shim layer here for specific capabilities should be fine IMO.

File: examples/multimodal/components/decode_worker.py (hunk 5 above)
Comment (krishung5) [reply examples/multimodal/components/decode_worker.py:161]: Right now it only takes bs=1. I think we can extend it to support batches in later phases for perf improvements.

**Comment by piotrm-nvidia** [comment]
I profiled the LLaVA multimodal inference pipeline with `timeit` and `%timeit` in ipython.


**Summary Table of Measured Steps**

| Step                   | Operation Description                     | Time      |
| ---------------------- | ----------------------------------------- | --------- |
| 1. Image Load          | HTTP GET + decode                         | \~2.1 ms  |
| 2. Image Preprocessing | Resize, normalize, tensorize              | \~4.15 ms |
| 3. Vision Tower        | ViT encoding of image                     | \~6.54 ms |
| 4. Project to MM Space | Linear projection for fusion              | \~0.07 ms |
| 5. Full Pipeline       | Text-image integration + decoding (LLaVA) | \~506 ms  |



```mermaid
gantt
    dateFormat  X
    axisFormat %M
    title LLaVA Multimodal Inference Pipeline Timeline

    section Image Preparation
    Download and Decode Image       :active, a1, 0, 2s
    Preprocess Image                :active, a2, after a1, 4s

    section Vision Encoder
    Vision Tower (ViT)             :active, a3, after a2, 6.5s
    Project to MM Embedding Space  :active, a4, after a3, 0.07s

    section Language Model (Decoder)
    Chat Template & Generation     :active, a5, after a4, 493s

```

**Observations**

* **Vision component** is relatively lightweight (\~13 ms end-to-end).
* **Language model decoding** dominates the cost (\~500 ms), even with `fp16` on GPU and relatively short generation (`max_new_tokens=30`).
* **Projection head** is extremely fast and negligible in cost.


**Profiling with timeit**

Below is the code I used to profile the pipeline.

```python
image = "http://images.cocodataset.org/test2017/000000155781.jpg"
response = requests.get(image)
image_data = Image.open(BytesIO(response.content)).convert("RGB")
```
2.1 ms ± 8.13 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


```
image_embeds = image_processor(images=image_data, return_tensors="pt")
```

4.15 ms ± 12.5 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


```
vision_outputs = vision_model.vision_tower(tower_inputs)
```    
6.54 ms ± 11.6 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)

```
embeddings = vision_model.multi_modal_projector(vision_outputs.last_hidden_state)
```
69.5 μs ± 280 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)



```python
import torch
from transformers import AutoProcessor, LlavaForConditionalGeneration

# Load the model in half-precision
model = LlavaForConditionalGeneration.from_pretrained("llava-hf/llava-1.5-7b-hf", torch_dtype=torch.float16, device_map="auto")
processor = AutoProcessor.from_pretrained("llava-hf/llava-1.5-7b-hf")

conversation = [
    {
        "role": "user",
        "content": [
            {"type": "image", "url": "/workspace/image/australia.jpg"},
            {"type": "text", "text": "What is shown in this image?"},
        ],
    },
]


# Generate
def generate(model, processor, conversation):
    inputs = processor.apply_chat_template(
        conversation,
        add_generation_prompt=True,
        tokenize=True,
        return_dict=True,
        return_tensors="pt"
    ).to(model.device, torch.float16)
    generate_ids = model.generate(**inputs, max_new_tokens=30)
    return processor.batch_decode(generate_ids, skip_special_tokens=True)

generate(model, processor, conversation)
```

```
In [43]: timeit generate(model, processor, conversation)
506 ms ± 1.31 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

File: examples/multimodal/components/encode_worker.py (hunk 6 above)
Comment (whoisj) [reply examples/multimodal/components/encode_worker.py]: Image caching has been added to the PR.

File: examples/multimodal/connect/__init__.py (hunk 3 above)
Comment (whoisj) [reply examples/multimodal/connect/__init__.py:1]: @mkhazraee and providing a "pythonic" interface with built-in documentation are important aspects here.

The overhead introduced by this layer is virtually non-existent compared to the data transfer times; and even the data transfer times are a small portion of the execution time.

Comment (mkhazraee) [reply examples/multimodal/connect/__init__.py:1]: Resolving this following our conversation.

File: examples/multimodal/components/decode_worker.py (hunk 1 above)
Comment (krishung5) [reply examples/multimodal/components/decode_worker.py:145]: Maybe can fix this as follow-up in a later phase, not a blocker for this PR. Could we add TODO comment for it?

File: examples/multimodal/connect/__init__.py (hunk 3 above)
Comment (krishung5) [reply examples/multimodal/connect/__init__.py:1]: @whoisj @nnshah1 Could you clarify what was the issue? I remembered it has something to do with circular dependency?

**Review by krishung5 — COMMENTED** [review]
Can we update the README.md as well to mention the NIXL transfer? I believe it says NATS.io right now. Overall LGTM, thanks!

**Comment by zhshw** [comment]
@krishung5 This PR submission is valuable. While the documentation update can be deferred, please merge it as soon as possible.

**Review by krishung5 — APPROVED** [review]

**Comment by krishung5** [comment]
@whoisj Lets add the documentation as follow-up. Ticket: DIS-98.
